### PR TITLE
New DB connection form

### LINKED ISSUE #3311: Front end changes for the "New Database Connection" form
See these [specs](https://wiki.mathesar.org/design/specs/new-db-connection-form/).

| Case A | Case B | Case C |
| -- | -- | -- |
|  |  |  |

### PR DESCRIPTION
**Pull request by seancolsen**
Fixes #3311

## Discrepancies from spec

This PR follows the [spec](https://wiki.mathesar.org/design/specs/new-db-connection-form/) except in these places:

### Connection nickname displays in connection selector

I decided to display the connection nickname to the connection selector dropdown:

| Spec | PR |
| -- | -- |
| image | image |

This makes the options easier to read and has the added benefit of giving us a easy way to distinguish the internal connection from the user database connections.

### "Learn more" docs hyperlinks missing

The spec has some "Learn more" hyperlinks inside help text which should point to docs pages. I'd like to add those in a subsequent PR, potentially even in a subsequent release. I'll open an issue for it after we merge this PR and we can discuss priority at that point.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by seancolsen** [comment]
@ghislaineguerin @kgodey @pavish @mathemancer since (_I think_) you were all involved in reviewing the [spec](https://wiki.mathesar.org/design/specs/new-db-connection-form/) I want to give you a heads up about a minor design change that I improvised within this PR. See the **"Connection nickname displays in connection selector"** section above and let me know if you have any concerns.

**Comment by seancolsen** [comment]
@pavish this PR is ready for review but is still marked draft due to being stacked on top of #3348. Since it's rather large (and also high priority), I'm assigning you to review it now so that we can hopefully get a head start. Make sure not to merge it until #3348 is merged though.

**Comment by kgodey** [comment]
> See the "Connection nickname displays in connection selector" section above and let me know if you have any concerns.

No concerns.

**Comment by pavish** [comment]
> "Connection nickname displays in connection selector"

No concerns from me.

**Comment by ghislaineguerin** [comment]
@seancolsen looks good. no concerns

File: mathesar_ui/src/systems/connections/AddConnection.svelte
Comment (pavish) [inline mathesar_ui/src/systems/connections/AddConnection.svelte:123]: @seancolsen 

This logic is making me question the entire UX for Case C (Creating a new Postgres user), as I review it. 
1. If we're going to check for the same host name and port with the list of connections at hand to enable creating a new user, why does the user need to input the host name and port instead of selecting from the list?
1. If the user is going to input the host and the port, why would they want to select an existing user account to create a new one for Case C? Wouldn't this make more sense with Case A?

I understand this should have been discussed earlier, but the UX is confusing as I attempt to use it, and we should discuss this once again. cc: @ghislaineguerin, @kgodey

Comment (pavish) [reply mathesar_ui/src/systems/connections/AddConnection.svelte:123]: Here's a screen recording describing the issue (Please enable audio).

https://github.com/mathesar-foundation/mathesar/assets/11850603/9e0af15d-ea52-40fd-92b7-3f58723dbc91

File: mathesar_ui/src/component-library/render/Render.svelte
Comment (seancolsen) [reply mathesar_ui/src/component-library/render/Render.svelte:17]: @pavish, it sounds like you're proposing we get rid of this component and inline its definition into all of its call sites. I'm open to that — if we can figure out a clean way to do it. This component is only called in one place right now, but I imagine using it more. Basically any time we have a `ComponentWithProps` instance, this component renders it in a type-safe way.

With this component, we do:

```html
<Render componentWithProps={foo}/>
```

Without this component, we would do:

```html
<svelte:component this={foo.component} {...foo.props} />
```

That's a bit more verbose, but not terrible. The main problem is this type error I'm getting when I try to do that:

image

I could be wrong, but this seems like a limitation of the typing provided by svelte-language-tools. It could be worth reporting this upstream to see what advice others have. I just didn't want to spend time on that right now.

Basically: wrapping the usage of `any` into a component seemed like a better approach to me because that way the dirty stuff is all wrapped up into one place that we can fix later on, even if we end up needing to use this in several places throughout the app.

Feel free to tinker with this and propose something better though.

File: mathesar_ui/src/systems/connections/AddConnection.svelte
Comment (seancolsen) [reply mathesar_ui/src/systems/connections/AddConnection.svelte:123]: We briefly [discussed](https://matrix.to/#/!kpLRqiyQwsndKhBIvd:matrix.mathesar.org/$elSq90BM9oH6HC3YeKwUgf_7GUkQ_WtD9bHylmSNZQw?via=matrix.mathesar.org&via=matrix.org) this on Matrix today. Next steps are for us to have more discussion after holiday break.

Comment (seancolsen) [reply mathesar_ui/src/systems/connections/AddConnection.svelte:123]: We discussed this today at our maintainers meeting and decided to remove case C entirely. I'll be working on that next.

Comment (seancolsen) [reply mathesar_ui/src/systems/connections/AddConnection.svelte]: I removed case C in e3d0ceeb504ffc63eae403f4deea21fc52345de3

**Comment by seancolsen** [comment]
@pavish this is ready for further review now.

File: mathesar_ui/src/component-library/render/Render.svelte
Comment (seancolsen) [reply mathesar_ui/src/component-library/render/Render.svelte]: Also I added c30b99f513910b1afadc0c67249effe2661b6610 which I hope cleans this up a bit more.

Comment (pavish) [reply mathesar_ui/src/component-library/render/Render.svelte]: > wrapping the usage of any into a component seemed like a better approach to me because that way the dirty stuff is all wrapped up into one place that we can fix later on, even if we end up needing to use this in several places throughout the app

Okay, this seems good for now. `svelte-language-tools` doesn't seem to work great with generics and SvelteComponentTyped. I would like for us to figure out a better solution for this, probably, after a svelte version upgrade.

**Review by pavish — APPROVED** [review]
Looks good!

**Pull request merged**
